coalib/core/Bear.py

```diff
@@ -0,0 +1,469 @@
+        else:
+            current_results = self._dependency_results[bear_type]
+
+        current_results += dependency_results
+
+    @property
+    def dependency_results(self):
+        """
+        Contains all dependency results.
+
+        This variable gets set during bear execution from the core and can be
+        used from ``analyze``.
+
+        Modifications to the returned dictionary lead to undefined behaviour.
+
+        >>> section = Section('my-section')
+        >>> file_dict = {'file1.txt': ['']}
+        >>> bear = Bear(section, file_dict)
+        >>> bear.dependency_results
+        {}
+        >>> dependency_bear = Bear(section, file_dict)
+        >>> bear.add_dependency_results(dependency_bear, [1, 2, 3])
+        >>> bear.dependency_results
+        {<class 'coalib.core.Bear.Bear'>: [1, 2, 3]}
+
+        :return:
+            A dictionary with bear-types as keys and their results received.
+        """
+        return self._dependency_results
+
+    @classmethod
+    def get_metadata(cls):
+        """
+        :return:
+            Metadata for the ``analyze`` function extracted from its signature.
+            Excludes parameter ``self``.
+        """
+        return FunctionMetadata.from_function(
+            cls.analyze,
+            omit={'self'})
+
+    # FIXME Make this a @classproperty.
+    @classmethod
+    def get_non_optional_settings(cls):
+        """
+        This method has to determine which settings are needed by this bear.
+        The user will be prompted for needed settings that are not available
+        in the settings file so don't include settings where a default value
+        would do.
+
+        :return: A dictionary of needed settings as keys and a tuple of help
+                 text and annotation as values
+        """
+        return cls.get_metadata().non_optional_params
+
+    @classmethod
+    def __json__(cls):
+        """
+        Override JSON export of ``Bear`` class.
+        """
+        # Those members get duplicated if they aren't excluded because they
+        # exist also as fields.
+        excluded_members = {'can_detect', 'maintainers', 'maintainers_emails'}
+
+        # json cannot serialize properties, so drop them
+        data = {
+            key: value
+            for key, value in get_public_members(cls).items()
+            if not isinstance(value, property) and key not in excluded_members}
+
+        metadata = cls.get_metadata()
+        non_optional_params = metadata.non_optional_params
+        optional_params = metadata.optional_params
+        data['metadata'] = {
+            'desc': metadata.desc,
+            'non_optional_params': {param: non_optional_params[param][0]
+                                    for param in non_optional_params},
+            'optional_params': {param: optional_params[param][0]
+                                for param in optional_params}}
+
+        return data
+
+    @staticmethod
+    def setup_dependencies():
```

Comment:
nope, the base bear has no dependencies, so the setup step does not do anything :)